flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/aggregate/RowTimeUnboundedOver.java

```diff
@@ -0,0 +1,266 @@
+					registerProcessingCleanupTimer(ctx, ctx.timerService().currentProcessingTime());
+				}
+			}
+			return;
+		}
+
+		Iterator<Map.Entry<Long, List<BaseRow>>> keyIterator = inputState.iterator();
+		if (keyIterator != null && keyIterator.hasNext()) {
+			Long curWatermark = ctx.timerService().currentWatermark();
+			Boolean existEarlyRecord = false;
+
+			// sort the record timestamps
+			do {
+				Long recordTime = keyIterator.next().getKey();
+				// only take timestamps smaller/equal to the watermark
+				if (recordTime <= curWatermark) {
+					insertToSortedList(recordTime);
+				} else {
+					existEarlyRecord = true;
+				}
+			} while (keyIterator.hasNext());
+
+			// get last accumulator
+			BaseRow lastAccumulator = accState.value();
+			if (lastAccumulator == null) {
+				// initialize accumulator
+				lastAccumulator = function.createAccumulators();
+			}
+			// set accumulator in function context first
+			function.setAccumulators(lastAccumulator);
+
+			// emit the rows in order
+			while (!sortedTimestamps.isEmpty()) {
+				Long curTimestamp = sortedTimestamps.removeFirst();
+				List<BaseRow> curRowList = inputState.get(curTimestamp);
+				if (curRowList != null) {
+					// process the same timestamp datas, the mechanism is different according ROWS or RANGE
+					processElementsWithSameTimestamp(curRowList, out);
+				} else {
+					// Ignore the same timestamp datas if the state is cleared already.
+					LOG.warn("The state is cleared because of state ttl. " +
+						"This will result in incorrect result. " +
+						"You can increase the state ttl to avoid this.");
+				}
+				inputState.remove(curTimestamp);
+			}
+
+			// update acc state
+			lastAccumulator = function.getAccumulators();
+			accState.update(lastAccumulator);
+
+			// if are are rows with timestamp > watermark, register a timer for the next watermark
+			if (existEarlyRecord) {
+				ctx.timerService().registerEventTimeTimer(curWatermark + 1);
+			}
+		}
+
+		// update cleanup timer
+		registerProcessingCleanupTimer(ctx, ctx.timerService().currentProcessingTime());
+	}
+
+	/**
+	 * Inserts timestamps in order into a linked list.
+	 * If timestamps arrive in order (as in case of using the RocksDB state backend) this is just
+	 * an append with O(1).
+	 */
+	private void insertToSortedList(Long recordTimestamp) {
+		ListIterator<Long> listIterator = sortedTimestamps.listIterator(sortedTimestamps.size());
+		Boolean isContinue = true;
+		while (listIterator.hasPrevious() && isContinue) {
+			Long timestamp = listIterator.previous();
+			if (recordTimestamp >= timestamp) {
+				listIterator.next();
+				listIterator.add(recordTimestamp);
+				isContinue = false;
+			}
+		}
+
+		if (isContinue) {
+			sortedTimestamps.addFirst(recordTimestamp);
+		}
+	}
+
+	/**
+	 * Process the same timestamp datas, the mechanism is different between
+	 * rows and range window.
+	 */
+	public void processElementsWithSameTimestamp(
```

Comment:
I think this method should be abstract, so that the sub-classes know to override it.